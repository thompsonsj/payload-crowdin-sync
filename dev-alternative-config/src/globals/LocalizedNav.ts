import type { GlobalConfig } from "payload";

export const LocalizedNav: GlobalConfig = {
  slug: "localized-nav",
  fields: [
    {
      name: "items",
      type: "array",
      required: true,
      maxRows: 8,
      fields: [
        {
          name: "label",
          type: "text",
          localized: true,
        },
      ],
    },
  ],
};

