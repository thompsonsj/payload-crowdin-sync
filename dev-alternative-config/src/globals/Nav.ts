import type { GlobalConfig } from "payload";

const Nav: GlobalConfig = {
  slug: "nav",
  fields: [
    {
      name: "items",
      type: "array",
      required: true,
      maxRows: 8,
      fields: [
        {
          name: "label",
          type: "text",
        },
      ],
    },
  ],
};

export default Nav;
