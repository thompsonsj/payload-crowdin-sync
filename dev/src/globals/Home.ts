/**
 * Home global
 * 
 * Test nested fields with nested localized props.
 */

import type { GlobalConfig } from "payload";

export const Home: GlobalConfig = {
  slug: "home",
  fields: [
    {
      type: "tabs",
      tabs: [
        {
          label: 'Navigation Cards 1',
          description: 'Product feature navigation cards.',
          fields: [
            {
              label: "Colapsible",
              type: "collapsible",
              fields: [
                {
                  name: 'parentGroup',
                  type: 'group',
                  localized: true,
                  fields: [
                    {
                      name: 'preTitle',
                      type: 'text',
                    },
                    {
                      name: 'title',
                      type: 'text',
                    },
                    {
                      name: 'subGroupOne',
                      type: 'group',
                      fields: [
                        {
                          name: 'title',
                          type: 'text',
                        },
                        {
                          name: 'text',
                          type: 'text',
                        },
                        {
                          name: 'ctaText',
                          type: 'text',
                        },
                        {
                          name: 'image',
                          type: 'upload',
                          relationTo: 'media',
                        },
                      ],
                    },
                    {
                      name: 'subGroupTwo',
                      type: 'group',
                      localized: true,
                      fields: [
                        {
                          name: 'title',
                          type: 'text',
                        },
                        {
                          name: 'text',
                          type: 'text',
                        },
                        {
                          name: 'ctaText',
                          type: 'text',
                        },
                        {
                          name: 'image',
                          type: 'upload',
                          relationTo: 'media',
                        },
                      ],
                    },
                  ],
                },
              ],
            }
          ],
        },
      ],
    },
  ],
};
