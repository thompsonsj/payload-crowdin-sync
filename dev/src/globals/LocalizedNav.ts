import type { GlobalConfig } from 'payload'

export const LocalizedNav: GlobalConfig = {
  slug: 'localized-nav',
  versions: {
    drafts: true,
  },
  fields: [
    {
      name: 'items',
      type: 'array',
      required: true,
      maxRows: 8,
      fields: [
        {
          name: 'label',
          type: 'text',
          localized: true,
        },
      ],
    },
  ],
}
