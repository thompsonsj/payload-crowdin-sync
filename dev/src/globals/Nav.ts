import { GlobalConfig } from "payload/types";

const Nav: GlobalConfig = {
  slug: "nav",
  fields: [
    {
      name: "items",
      type: "array",
      required: true,
      maxRows: 8,
      fields: [
        {
          name: "label",
          type: "text",
          localized: true,
        },
      ],
    },
  ],
};

export default Nav;
