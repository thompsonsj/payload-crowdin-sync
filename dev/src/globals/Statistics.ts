import { GlobalConfig } from 'payload/types';

const Statistics: GlobalConfig = {
  slug: 'statistics',
  access: {
    read: () => true,
  },
  admin: {
    group: 'Shared',
  },
  versions: {
    drafts: true,
  },
  fields: [
    {
      name: 'users',
      type: 'group',
      fields: [
        {
          name: 'text',
          type: 'text',
          localized: true,
        },
        {
          name: 'number',
          type: 'number',
          min: 0,
          admin: {
            step: 100,
            description:
              'Restricted to multiples of 100 in order to simplify localization.',
          },
        },
      ],
    },
    {
      name: 'countries',
      type: 'group',
      fields: [
        {
          name: 'text',
          type: 'text',
          localized: true,
        },
        {
          name: 'number',
          type: 'number',
          min: 0,
          admin: {
            step: 1,
          },
        },
      ],
    },
  ],
};

export default Statistics;
