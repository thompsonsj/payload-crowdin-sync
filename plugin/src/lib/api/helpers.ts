import payload, { Payload } from "payload";
import { CrowdinArticleDirectory, CrowdinCollectionDirectory, CrowdinFile } from "../payload-types";
import { payloadCrowdinSyncTranslationsApi } from "./payload-crowdin-sync/translations";
import { PluginOptions, isCollectionOrGlobalConfigObject, isCollectionOrGlobalConfigSlug } from "../types";
import { ReadableStreamDefaultController } from "stream/web";
import { Plugin } from "payload/config";

/**
 * get Crowdin Article Directory for a given documentId
 *
 * The Crowdin Article Directory is associated with a document,
 * so is easy to retrieve. Use this function when you only have
 * a document id.
 */
export async function getArticleDirectory(
  documentId: string,
  payload: Payload,
  allowEmpty?: boolean
) {
  // Get directory
  const crowdinPayloadArticleDirectory = await payload.find({
    collection: "crowdin-article-directories",
    where: {
      name: {
        equals: documentId,
      },
    },
  });
  if (crowdinPayloadArticleDirectory.totalDocs === 0 && allowEmpty) {
    // a thrown error won't be reported in an api call, so console.log it as well.
    console.log(`No article directory found for document ${documentId}`);
    throw new Error(
      "This article does not have a corresponding entry in the  crowdin-article-directories collection."
    );
  }
  return crowdinPayloadArticleDirectory
    ? crowdinPayloadArticleDirectory.docs[0]
    : undefined;
}

export async function getFile(
  name: string,
  crowdinArticleDirectoryId: string,
  payload: Payload
): Promise<any> {
  const result = await payload.find({
    collection: "crowdin-files",
    where: {
      field: { equals: name },
      crowdinArticleDirectory: {
        equals: crowdinArticleDirectoryId,
      },
    },
  });
  return result.docs[0];
}

export async function getFiles(
  crowdinArticleDirectoryId: string,
  payload: Payload
): Promise<any> {
  const result = await payload.find({
    collection: "crowdin-files",
    limit: 10000,
    where: {
      crowdinArticleDirectory: {
        equals: crowdinArticleDirectoryId,
      },
    },
  });
  return result.docs;
}

export async function getFileByDocumentID(
  name: string,
  documentId: string,
  payload: Payload
): Promise<CrowdinFile> {
  const articleDirectory = await getArticleDirectory(documentId, payload);
  return getFile(name, `${articleDirectory?.id}`, payload);
}

export async function getFilesByDocumentID(
  documentId: string,
  payload: Payload
): Promise<CrowdinFile[]> {
  const articleDirectory = await getArticleDirectory(documentId, payload);
  if (!articleDirectory) {
    // tests call this function to make sure files are deleted
    return [];
  }
  const files = await getFiles(`${articleDirectory.id}`, payload);
  return files;
}

interface IupdatePayloadTranslation {
  articleDirectoryId: string
  pluginOptions: PluginOptions,
  payload: Payload
  /** store translations into a draft */
  draft?: boolean
  /** prevent database changes */
  dryRun?: boolean
  /** override article directory exclude locales */
  excludeLocales?: string[]
}

export async function updatePayloadTranslation({
  articleDirectoryId,
  pluginOptions,
  payload,
  draft,
  dryRun,
  excludeLocales,
}: IupdatePayloadTranslation) {
  // get article directory
  const articleDirectory = await payload.findByID({
    id: articleDirectoryId,
    collection: "crowdin-article-directories",
  });
  // is this a global or a collection?
  const global =
    (articleDirectory['crowdinCollectionDirectory'] as CrowdinCollectionDirectory)?.collectionSlug as string === "globals";
  // get an instance of our translations api
  const translationsApi = new payloadCrowdinSyncTranslationsApi(
    pluginOptions,
    payload
  );
  try {
    const translations = await translationsApi.updateTranslation({
      documentId: !global ? articleDirectory["name"] as string : ``,
      collection: global
        ? articleDirectory['name'] as string
        : (articleDirectory["crowdinCollectionDirectory"] as CrowdinCollectionDirectory)?.collectionSlug as string,
      global,
      draft,
      dryRun,
      excludeLocales: excludeLocales || articleDirectory["excludeLocales"] as string[] || [],
    });
    return {
      status: 200,
      ...translations
    };
  } catch (error) {
    return {
      status: 400,
      error,
    }
  }
}

export const isCrowdinActive = ({
  doc,
  slug,
  global,
  pluginOptions
}: {
  doc: any,
  slug: string
  global: boolean
  pluginOptions: PluginOptions
}) => {
  if (!slug) {
    return false
  }

  if (global) {
    if (!pluginOptions.globals) {
      return true
    }

    const matchingGlobalConfig = pluginOptions.globals.find(config => {
      if (isCollectionOrGlobalConfigObject(config) && config.slug === slug) {
        if (typeof config.condition !== 'undefined') {
          return config.condition({ doc })
        }
        return true
      }
      if (isCollectionOrGlobalConfigSlug(config)) {
        return config === slug
      }
      return false
    })

    if (typeof matchingGlobalConfig !== 'undefined') {
      return true
    }
  }

  if (!pluginOptions.collections) {
    return true
  }

  const matchingCollectionConfig = pluginOptions.collections.find(config => {
    if (isCollectionOrGlobalConfigObject(config) && config.slug === slug) {
      if (typeof config.condition !== 'undefined') {
        return config.condition({ doc })
      }
      return true
    }
    if (isCollectionOrGlobalConfigSlug(config)) {
      return config === slug
    }
    return false
  })

  if (typeof matchingCollectionConfig !== 'undefined') {
    return true
  }

  return false
}
