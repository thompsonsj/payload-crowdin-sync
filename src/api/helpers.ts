import { Payload } from "payload";
import { IcrowdinFile } from "./payload-crowdin-sync/files";

/**
 * get Crowdin Article Directory for a given documentId
 *
 * The Crowdin Article Directory is associated with a document,
 * so is easy to retrieve. Use this function when you only have
 * a document id.
 */
export async function getArticleDirectory(
  documentId: string,
  payload: Payload,
  allowEmpty?: boolean
) {
  // Get directory
  const crowdinPayloadArticleDirectory = await payload.find({
    collection: "crowdin-article-directories",
    where: {
      name: {
        equals: documentId,
      },
    },
  });
  if (crowdinPayloadArticleDirectory.totalDocs === 0 && allowEmpty) {
    // a thrown error won't be reported in an api call, so console.log it as well.
    console.log(`No article directory found for document ${documentId}`);
    throw new Error(
      "This article does not have a corresponding entry in the  crowdin-article-directories collection."
    );
  }
  return crowdinPayloadArticleDirectory
    ? crowdinPayloadArticleDirectory.docs[0]
    : undefined;
}

export async function getFile(
  name: string,
  crowdinArticleDirectoryId: string,
  payload: Payload
): Promise<any> {
  const result = await payload.find({
    collection: "crowdin-files",
    where: {
      field: { equals: name },
      crowdinArticleDirectory: {
        equals: crowdinArticleDirectoryId,
      },
    },
  });
  return result.docs[0];
}

export async function getFiles(
  crowdinArticleDirectoryId: string,
  payload: Payload
): Promise<any> {
  const result = await payload.find({
    collection: "crowdin-files",
    limit: 10000,
    where: {
      crowdinArticleDirectory: {
        equals: crowdinArticleDirectoryId,
      },
    },
  });
  return result.docs;
}

export async function getFileByDocumentID(
  name: string,
  documentId: string,
  payload: Payload
): Promise<IcrowdinFile> {
  const articleDirectory = await getArticleDirectory(documentId, payload);
  return getFile(name, `${articleDirectory?.id}`, payload);
}

export async function getFilesByDocumentID(
  documentId: string,
  payload: Payload
): Promise<IcrowdinFile[]> {
  const articleDirectory = await getArticleDirectory(documentId, payload);
  if (!articleDirectory) {
    // tests call this function to make sure files are deleted
    return [];
  }
  const files = await getFiles(`${articleDirectory.id}`, payload);
  return files;
}
