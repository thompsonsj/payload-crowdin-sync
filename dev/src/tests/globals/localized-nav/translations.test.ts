import {
    // payloadCrowdinSyncTranslationsApi,
    utilities,
    mockCrowdinClient,
    // getFilesByDocumentID,
    // getFiles,
  } from 'payload-crowdin-sync'
  import { fixtures } from './fixtures'
  import nock from 'nock'
  
  import { pluginConfig } from '../../helpers/plugin-config'
  import {
    CrowdinArticleDirectory,
    LocalizedNav,
  } from '@/payload-types'
  
  export const isNotString = <T>(val: T | string | undefined | null): val is T => {
    return val !== undefined && val !== null && typeof val !== 'string'
  }
  
  const getRelationshipId = (relationship?: string | CrowdinArticleDirectory | null) => {
    if (!relationship) {
      return undefined
    }
    if (isNotString(relationship)) {
      return relationship.id
    }
    return relationship
  }
  
  import { initPayloadInt } from '../../helpers/initPayloadInt'
  import type { Payload } from 'payload'
import { LocalizedNav as LocalizedNavConfig } from './../../../globals/LocalizedNav'
  
  let payload: Payload
  
  const pluginOptions = pluginConfig()
  const mockClient = mockCrowdinClient(pluginOptions)
  
  describe('Global: localized-nav', () => {
    beforeAll(async () => {
      const initialized = await initPayloadInt()
      ;({ payload } = initialized as {
        payload: Payload
      })
    })
  
    afterEach((done) => {
      if (!nock.isDone()) {
        throw new Error(`Not all nock interceptors were used: ${JSON.stringify(nock.pendingMocks())}`)
      }
      nock.cleanAll()
      done()
    })
  
    afterAll(async () => {
      if (typeof payload.db.destroy === 'function') {
        await payload.db.destroy()
      }
    })

  
    it('builds a Crowdin JSON object as expected', async () => {
      nock('https://api.crowdin.com')
        .post(`/api/v2/projects/${pluginOptions.projectId}/directories`)
        .times(2)
        .reply(200, mockClient.createDirectory({}))
        .post(`/api/v2/storages`)
        .times(1)
        .reply(200, mockClient.addStorage())
        .post(`/api/v2/projects/${pluginOptions.projectId}/files`)
        .times(1)
        .reply(200, mockClient.createFile({}))
  
      const global = (await payload.updateGlobal({
        slug: 'localized-nav',
        data: {
          items: fixtures.published.items,
        },
      })) as LocalizedNav
  
      expect(
        utilities.buildCrowdinJsonObject({
          doc: global,
          fields: LocalizedNavConfig.fields,
        }),
      ).toMatchSnapshot()
    })
  })
  