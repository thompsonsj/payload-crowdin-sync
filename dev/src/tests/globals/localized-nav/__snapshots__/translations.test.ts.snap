// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Global: localized-nav builds a Crowdin JSON object as expected 1`] = `
{
  "items": {
    "68ddb03a9432e06617d52103": {
      "label": "Published nav item 1",
    },
    "68ddb04d9432e06617d52105": {
      "label": "Published nav item 2",
    },
  },
}
`;
